reference/src/representation/enums.md

```diff
@@ -0,0 +1,376 @@
+"C" fn()>` will be represented at runtime as a nullable function
+pointer (which is therefore equivalent to a C function pointer) . FFI
+interop often depends on this property.
+
+**Example.** Consider the following enum definitions:
+
+```rust
+enum Enum1<T> {
+  Present(T),
+  Absent1,
+  Absent2,
+}
+
+enum Enum2 {
+  A, B, C
+}  
+```
+
+`Enum1<&u8>` is not eligible for discriminant elision, since `&u8`
+defines a single niche value, but `Enum1` has two unit
+variants. However, `Enum2` has only three legal values (0 for `A`, 1
+for `B`, and 2 for `C`), and hence defines a plethora of niche values[^caveat].
+Therefore, `Enum1<Enum2>` is guaranteed to be laid out the same as
+`Enum2` ([consider the results of applying
+`size_of`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=eadff247f2c5713b8f3b6c9cda297711)).
+
+[^caveat]: Strictly speaking, niche values are considered part of the "representation invariant" for an enum and not its type. Therefore, this section is added only as a preview for future unsafe-code-guidelines discussion.
+
+### Other optimizations
+
+The previous section specified a relatively narrow set of layout
+optimizations that are **guaranteed** by the compiler. However, the
+compiler is always free to perform **more** optimizations than this
+minimal set. For example, the compiler presently treats `Result<T,
+()>` and `Option<T>` as equivalent, but this behavior is not
+guaranteed to continue as `Result<T, ()>` is not considered
+"option-like".
+
+As of this writing, the compiler's current behavior is to attempt to
+elide discriminants whenever possible. Furthermore, a variant whose
+only fields are of zero-size is considered a unit variant for this
+purpose. If eliding discriminants is not possible (e.g., because the
+payload does not define sufficient niche values), then the compiler
+will select an appropriate discriminant size `N` and use a
+representation roughly equivalent to `#[repr(N)]`, though without the
+strict `#[repr(C)]` guarantees on each struct. However, this behavior
+is not guaranteed to remain the same in future versions of the
+compiler and should not be relied upon. (While it is not expected that
+existing layout optimizations will be removed, it is possible -- it is
+also possible for the compiler to introduce new sorts of
+optimizations.)
+
+## Niche values
+
+C-like enums with N variants and no specified representation are
+guaranteed to supply niche values corresponding to 256 - N (presuming
+that is a positive number). This is because a C-like enum must be
+represented using an integer and that integer must correspond to a
+valid variant: the precise size of C-like enums is not specified but
+it must be at least one byte, which means that there are at least 256
+possible bitstrings (only N of which are valid).
+
+Other enums -- or enums with a specified representation -- may supply
+niches if their representation invariant permits it, but that is not
+**guaranteed**.
```

Comment:
It seems odd to have a partial defn of "niche" here.

Reply:
I removed that section regardless. but I'm not really sure what you meant by "partial" -- i.e., just covering enums? I was assuming each type would define for itself when it offers a niche. But we'll cover it later I guess.